---
title: "Next_steps"
output: html_document
author: "Ethan Ashby"
contact: "ecadeashby@gmail.com"
---

This file includes major next steps for future research.

My major goal in terms of directly addressing biological questions is to rewrite & tweak the method developed in this paper (https://academic.oup.com/bioinformatics/article/27/13/i392/183867), which integrates optimal clustering and impulse modeling to demonstrate & compare transcriptomic responses to stimuli. This method provides a number of advantages: clustering to models prevents clustering to noise (a real danger w/ RNA-Seq data). This method also helps solves the problem of the choice of the number of clusters, as it learns and develops optimal cluster groups. This is EXACTLY the kind of thing Profs Stoebel and Schulz want, but it is going to take a little (probably a lot) of work to do. Their documentation is pretty poor and the implementation is through Matlab (ew), and runs on a computing cluster (which I'm not sure if we have access to @ Pomona). If we could tweak some aspects (e.g. parameter fitting procedure) of their approach and rewrite the package in R, this would be an awesome resource for our Bio PIs. Alternatively, there are packages that allow you to integrate matlab code in R, but this would be less reproducible and less desireable.

My second major goal is more statistically focused: I would like to continue evaluating these models (sigmoid and impulse) to understand them better. When do these models represent the data well and when do they fail? I'd like to evaluate and quantify the robustness/stability of these models to dispersion based noise, outliers, number of data points, number of replicates. I also would like to settle on a method to quantify the GOF of each model to the data at hand. This may be very difficult, but perhaps we could even develop a tool/model over a bunch of TC RNA-seq datasets, which helps guide researchers in the design of their TC experiments (i.e. researcher provides the desired temporal resolution & number of samples, and the program outputs how many time points and replicates they should use).


