---
title: "PAM_tb_shulz"
author: "Lian Morales"
date: "7/9/2019"
output: html_document
---

```{r}
#LOADING DESEQ2 AND RNASEQ FILE
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


library("dplyr")
library("tidyr")
library("DESeq2")
library(readr)
library(devtools)
library("DESeq2")
library(dplyr)
library(readr)
library(tidyr)
library(ggplot2)
library("vsn")
library("hexbin")
library("corrplot")

#for GSEA
library(doParallel)
library(Biobase)
library(biomaRt)
library(fgsea)
```


```{r}
#DESEQ ON SCHULZ
library(readr)
TBcounts <- read.table("18-5-23_Proc_ibet_TC_for_deseq.txt",header = TRUE) %>% dplyr ::select(-ib_3h_1) 
TBcounts <- TBcounts %>% dplyr :: distinct(gene, .keep_all = TRUE) 
#NORMALIZATION BY DESEQ2
rownames(TBcounts) <- TBcounts$gene #renames the gene row
TBcounts <- TBcounts %>% dplyr :: select(-"gene") #this gets rid of the name gene in table
#creating new data frame for DESeq to analyze
tb_new <- data.frame(rownames = colnames(TBcounts),
                     reps = as.factor(c(1,2,3,1,2,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3)),
                    stage = as.factor(c(rep("noo_ib",3), rep("ib_3h", 2), rep("ib_6h",3), rep("ib_12h",3),rep("ib_24h",3),rep("ib_48h",3),rep("ib_3d",3),rep("ib_7d",3),rep("ib_10d",3),rep("ib_14d",3))))
rownames(tb_new) <- tb_new$rownames
tb_new <- tb_new %>% dplyr ::select(-rownames)

##dds stands for the DESeq Data Set
TBcounts <- TBcounts[rowSums(TBcounts) > 0 ,]
ddsFullCountTable <- DESeqDataSetFromMatrix(countData = TBcounts,
                                            colData = tb_new, 
                                            design = ~ reps + stage)

TBcounts_dds <- DESeq(ddsFullCountTable)

#calculating norm counts
##DESeq2 :: allows to specify where this fcn comes from
normDE_counts <- DESeq2::counts(TBcounts_dds, normalized = TRUE) 
normDE_counts_df <- data.frame(normDE_counts)
 

#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("ImpulseDE2")
#library(ImpulseDE2)

#time <- c(rep(0,3), rep(3,2), rep(6,3,), rep(12,3), rep(24,3), rep(48,3), rep(72,3), rep(168,3), rep(240,3), rep(336,3))
#impulse_design <- data.frame(Sample = colnames(TBcounts), Condition = rep("case", 29), Time = time)
#TBcounts <- as.matrix(TBcounts)
#tryp_impulse <- ImpulseDE2 :: runImpulseDE2(TBcounts, impulse_design,boolCaseCtrl = FALSE, vecConfounders = NULL, scaNProc = 8, scaQThres = 0.01)

```


```{r}
#COMPARING - DE ANALYSIS - FOR GSEA

#comparing stages noo_ib and ib_12h --- the stage of non inhibited bromodomain at time 0 to inhibited bromodomain at time point 12 hours
TBcountsRes <- results( TBcounts_dds ,contrast = c("stage","ib_12h", "noo_ib"))


#resSig stands for the results of the signigicant genes expressed in comparison of noo_ib and ib_6h
resSig <- resSig[complete.cases(resSig),]
resSig <- TBcountsRes[ which(TBcountsRes$padj < 0.01 ), ]
resSig$gene <- rownames(resSig)
resSig$fcsign <- sign(resSig$log2FoldChange) #fcsign stands for the log2foldchange sign

#creating negative and positive logfold change data sets
resSig_log.pos <- resSig[which(resSig$fcsign != -1),]
resSig_log.neg <- resSig[which(resSig$fcsign == -1),]

#GSEA -- MAKING DATA TABLES AND RUNNING FGSEA
  #Making metric rank for 0-12h comparison

#creating metric used for ranking DE of genes for TB schulz 0-12h timepoint contrast
#resSig$fcsign <- sign(resSig$log2FoldChange) #fcsign stands for the log2foldchange sign
#resSig <-  resSig[which(abs(resSig$fcsign) != 0),] #filtering out foldchanges = 0
resSig$logP = -log10(resSig$pvalue) 
resSig <-  resSig[which(abs(resSig$logP) != 0),] 
resSig$metric= resSig$logP * abs(resSig$log2FoldChange)#calculating metric = log10(pvalue)/fcsign
resSig$Gene <- rownames(resSig) #creating column Gene for dataframe
gsea_schulz <-resSig[,c("Gene", "metric")]

#Creating data frame from gsea_schulz for gsea
gsea_schulz_df <- data.frame(gsea_schulz)
rownames(gsea_schulz_df) <- NULL
gsea_schulz_df <- gsea_schulz_df[which(gsea_schulz_df$metric != -Inf),] #removing metric where = -Inf
gsea_schulz_df <- gsea_schulz_df[which(gsea_schulz_df$metric != Inf),]
 
#creating stats and pathways for gsea
gseaDat_schulz <- filter(gsea_schulz_df, !is.na(Gene))
ranks <- gsea_schulz_df$metric
names(ranks) <- gseaDat_schulz$Gene
ranks <- sort(ranks)


#####FOR LOG POS FOLD CHANGE######
#CREATING RANKS AND DATA FRAME FOR GSEA
#resSig_log.pos$fcsign <- sign(resSig_log.pos$log2FoldChange) #fcsign stands for the log2foldchange sign
#resSig_log.pos <-  resSig_log.pos[which(abs(resSig_log.pos$fcsign) != 0),] #filtering out foldchanges = 0
resSig_log.pos$logP = -log10(resSig_log.pos$pvalue) 
resSig_log.pos <-  resSig_log.pos[which(abs(resSig_log.pos$logP) != 0),] 
resSig_log.pos$metric= resSig_log.pos$logP * abs(resSig_log.pos$log2FoldChange) #metric = -log10(pvalue) *abs(log2FoldChange)
resSig_log.pos$Gene <- rownames(resSig_log.pos) #creating column Gene for dataframe
gsea_schulz_log.pos <-resSig_log.pos[,c("Gene", "metric")]
 
  #creating data frame
gsea_schulz_df_log.pos <- data.frame(gsea_schulz_log.pos)
rownames(gsea_schulz_df_log.pos) <- NULL
gsea_schulz_df_log.pos <- gsea_schulz_df_log.pos[which(gsea_schulz_df_log.pos$metric != -Inf),] #removing metric where = -Inf
gsea_schulz_df_log.pos <- gsea_schulz_df_log.pos[which(gsea_schulz_df_log.pos$metric != Inf),]

  #creating ranks for fgsea
gseaDat_schulz_log.pos <- filter(gsea_schulz_df_log.pos, !is.na(Gene))
ranks_log.pos <- gsea_schulz_df_log.pos$metric
names(ranks_log.pos) <- gseaDat_schulz_log.pos$Gene
ranks_log.pos <- sort(ranks_log.pos)

#####FOR LOG NEG FOLD CHANGE######
#CREATING RANKS AND DATA FRAME FOR GSEA
#resSig_log.neg$fcsign <- sign(resSig_log.neg$log2FoldChange) #fcsign stands for the log2foldchange sign
#resSig_log.neg <-  resSig_log.neg[which(abs(resSig_log.neg$fcsign) != 0),] #filtering out foldchanges = 0
resSig_log.neg$logP <-  -log10(resSig_log.neg$pvalue)
resSig_log.neg <-  resSig_log.neg[which(abs(resSig_log.neg$logP) != 0),] 
resSig_log.neg$metric <-  resSig_log.neg$logP * abs(resSig_log.neg$log2FoldChange) #calculating metric = log10(pvalue) 
resSig_log.neg$Gene <- rownames(resSig_log.neg) #creating column Gene for dataframe
gsea_schulz_log.neg <-resSig_log.neg[,c("Gene", "metric")]

  #creating data frame
gsea_schulz_df_log.neg <- data.frame(gsea_schulz_log.neg)
rownames(gsea_schulz_df_log.neg) <- NULL
gsea_schulz_df_log.neg <- gsea_schulz_df_log.neg[which(gsea_schulz_df_log.neg$metric != -Inf),] #removing metric   where = -Inf
gsea_schulz_df_log.neg <- gsea_schulz_df_log.neg[which(gsea_schulz_df_log.neg$metric != Inf),] 
  #creating ranks for fgsea
gseaDat_schulz_log.neg <- filter(gsea_schulz_df_log.neg, !is.na(Gene))
ranks_log.neg <- gsea_schulz_df_log.neg$metric
names(ranks_log.neg) <- gseaDat_schulz_log.neg$Gene
ranks_log.neg <- sort(ranks_log.neg)


```


```{r, include=FALSE}
#DESEQ ON SAVAGE DATA - THREE COMPARISON MGSA MGPV PVSA

savagecounts <- read.table("18-5-24_dedup_Savage_deseq_counts.txt", header = TRUE)

 
#Creating Data matrix 
rownames(savagecounts) <- savagecounts$gene #renames the gene row
savagecounts <- savagecounts %>% dplyr :: select(-"gene") #this gets rid of the name gene in table
coldata <- data.frame(rownames = colnames(savagecounts),
                    reps = as.factor(c(1, 2, 3, 1, 2, 3 ,4,
                                     1, 2 ,3)), 
                    stage = as.factor(c(rep("MG",3),rep("PV",4), rep("SA",3))))

ddsFullCountTable <- DESeqDataSetFromMatrix(
   countData = savagecounts,
   colData = coldata,
   design = ~ reps + stage)
#DESeq2 getting DE genes
savagecounts_dds <- DESeq(ddsFullCountTable )
```

```{r}
########CREATING GSEA LISTS FROM PATHWAY ANALYSIS OF SAV DE GENES
#Comparisons of all 
  #padj filter < 0.01
  #abs(log2fold) filter > 1.5

#MG and SA############################################################################
savageRes_MgSa <- results( savagecounts_dds ,contrast = c("stage",  "MG","SA"))
resSig <- savageRes_MgSa[ which(savageRes_MgSa$padj < 0.01 ), ]
#resSig <- resSig[which(abs(resSig$log2FoldChange) > 1.5),]
MgSa <- resSig
head(MgSa) 
###here we need to preform GO on these genes. Upload to tryp database under New search : metabolic pathways, and enter this txt file. Then, make sure to get is downloaded as a csv.
  #GO/PATHWAY ANALYSIS
 file_to_write<-paste(rownames(MgSa), collapse="\t")
write.table(file_to_write, file="MgSa_gene") #look up in file directory, then delete quotation marks as well as the first rows that are not genes


#MG and PV############################################################################
savageRes_MgPv <- results( savagecounts_dds ,contrast = c("stage",  "MG","PV"))
resSig <- savageRes_MgPv[ which(savageRes_MgPv$padj < 0.01 ), ]
#resSig <- resSig[which(abs(resSig$log2FoldChange) > 1.5),]
MgPv <- resSig

#creating file for GO/PATHWAY ANALYSIS
file_to_write<-paste(rownames(MgPv), collapse="\t")
write.table(file_to_write, file="MgPv_gene")

#PV and SA############################################################################
savageRes_PvSa <- results( savagecounts_dds ,contrast = c("stage",  "PV","SA"))
resSig <- savageRes_PvSa[ which(savageRes_PvSa$padj < 0.01 ), ]
#resSig <- resSig[which(abs(resSig$log2FoldChange) > 1.5),]
PvSa <- resSig

#creating file for GO/PATHWAY ANALYSIS
 file_to_write<-paste(rownames(PvSa), collapse="\t")
write.table(file_to_write, file="PvSa_gene")


#######MAKING RANKED METRIC LIST FOR SAVAGE DATA########

#making big data set 'allcomp' with all comparisons to make code easier. Splitting after by 'comparison' 
MGSA_df <- data.frame(Gene = rownames(MgSa), pvalue = MgSa$pvalue, padj = MgSa$padj, comparison = "MgSa",log2FoldChange= MgSa$log2FoldChange)
MGPV_df <- data.frame(Gene = rownames(MgPv), pvalue = MgPv$pvalue, padj = MgPv$padj, comparison = "MgPv", log2FoldChange= MgPv$log2FoldChange)
PVSA_df <- data.frame(Gene = rownames(PvSa), pvalue = PvSa$pvalue, padj = PvSa$padj, comparison = "PvSa",log2FoldChange= PvSa$log2FoldChange)
 
sav_allcomp <- rbind(MGSA_df,MGPV_df, PVSA_df)

#View(gsea_allcomp)

sav_allcomp$logP <- -log10(sav_allcomp$pvalue) #calculating metric == -log10(pvalue)
sav_allcomp <-  sav_allcomp[which(abs(sav_allcomp$logP) != 0),] #filtering out p-values 
sav_allcomp$metric <- sav_allcomp$logP * abs(sav_allcomp$log2FoldChange)
sav_gsea_allcomp <- sav_allcomp[,c("Gene","metric","comparison")]

MgSa$logP <- -log10(MgSa$pvalue)
MgSa$Gene <- rownames(MgSa)
MgSa <- MgSa[which(abs(MgSa$logP) != 0),]
MgSa$metric <- MgSa$logP * abs(MgSa$log2FoldChange)
MgSa_gsea  <- MgSa[,c("Gene","metric")]
gsea_MgSa_df <- data.frame(MgSa_gsea)
gseaDat_MgSa <- filter(gsea_MgSa_df, !is.na(Gene))
ranks_MgSa <- gsea_MgSa_df$metric
names(ranks_MgSa) <- gseaDat_MgSa$Gene
#ranks_MgSa <- sort(ranks_MgSa)
View(ranks_MgSa)
##filtering out big dataset -- creating data frames for each comparison 
gsea_MgSa_df<- sav_gsea_allcomp %>% filter(comparison == "MgSa") %>% dplyr :: select(-comparison) 
gsea_MgPv_df <- sav_gsea_allcomp %>% filter(comparison == "MgPv") %>% dplyr :: select(-comparison) 
gsea_PvSa_df <- sav_gsea_allcomp %>% filter(comparison == "PvSa") %>% dplyr :: select(-comparison) 
 
#creating stats and pathways for gsea
  #MgSa
gseaDat_MgSa <- filter(gsea_MgSa_df, !is.na(Gene))
ranks_MgSa <- gsea_MgSa_df$metric
names(ranks_MgSa) <- gseaDat_MgSa$Gene
ranks_MgSa <- sort(ranks_MgSa)
  #MgPv
gseaDat_MgPv <- filter(gsea_MgPv_df, !is.na(Gene))
ranks_MgPv <- gsea_MgPv_df$metric
names(ranks_MgPv) <- gseaDat_MgPv$Gene
ranks_MgPv <- sort(ranks_MgPv)
#PvSa
gseaDat_PvSa <- filter(gsea_PvSa_df, !is.na(Gene))
ranks_PvSa <- gsea_PvSa_df$metric
names(ranks_PvSa) <- gseaDat_PvSa$Gene
ranks_PvSa <- sort(ranks_PvSa)

```

```{r}

###############MAKING PATHWAY LIST FROM GO ANALYSIS - BIO, CELL,MOLECULAR###########################

####################BIO PROS GO###########################
MgSa_GObiopros <- read.table("MgSa_GObiopros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgSa_GObiopros <- data.frame(MgSa_GObiopros, comparison = "MgSa_bio")
MgPv_GObiopros <- read.table("MgPv_GObiopros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgPv_GObiopros <- data.frame(MgPv_GObiopros, comparison = "MgPv_bio")
PvSa_GObiopros <- read.table("PvSa_GObiopros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  PvSa_GObiopros <- data.frame(PvSa_GObiopros, comparison = "PvSa_bio")
  
all_bio <- rbind(MgSa_GObiopros,MgPv_GObiopros,PvSa_GObiopros)
all_bio$Result.gene.list <- gsub(',',',',all_bio$Result.gene.list,fixed=TRUE)
all_bio$Result.gene.list <- strsplit(all_bio$Result.gene.list, ",")
 
#MgSa Bio
MgSa_GObiopros <- data.frame(all_bio %>% dplyr :: filter(comparison == "MgSa_bio"))
MgSa_GObiopros_list <- as.list(MgSa_GObiopros)
names(MgSa_GObiopros_list$Result.gene.list) <- MgSa_GObiopros_list$Name
MgSaGOlist_bio <- MgSa_GObiopros_list$Result.gene.list
#MgPv Bio
MgPv_GObiopros <- data.frame(all_bio %>% filter(comparison == "MgPv_bio"))
MgPv_GObiopros_list <- as.list(MgPv_GObiopros)
names(MgPv_GObiopros_list$Result.gene.list) <- MgPv_GObiopros_list$Name
MgPvGOlist_bio <- MgPv_GObiopros_list$Result.gene.list
#PvSa Bio
PvSa_GObiopros <- data.frame(all_bio %>% filter(comparison == "PvSa_bio"))
PvSa_GObiopros_list <- as.list(PvSa_GObiopros)
names(PvSa_GObiopros_list$Result.gene.list) <- PvSa_GObiopros_list$Name
PvSaGOlist_bio <- PvSa_GObiopros_list$Result.gene.list


#####################CELL PROS#######################
  #MGSA
MgSa_GOcellpros <- read.table("MgSa_GOcellpros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgSa_GOcellpros <- data.frame(MgSa_GOcellpros, comparison = "MgSa_cell")
MgPv_GOcellpros <- read.table("MgPv_GOcellularpros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgPv_GOcellpros <- data.frame(MgPv_GOcellpros, comparison = "MgPv_cell")
PvSa_GOcellpros <- read.table("PvSa_GOcellularpros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  PvSa_GOcellpros <- data.frame(PvSa_GOcellpros, comparison = "PvSa_cell")
  
all_cell <- rbind(MgSa_GOcellpros, MgPv_GOcellpros,PvSa_GOcellpros)
all_cell$Result.gene.list <- gsub(',',',',all_cell$Result.gene.list,fixed=TRUE)
all_cell$Result.gene.list <- strsplit(all_cell$Result.gene.list, ",")

  #MGSA CELL
MgSa_GOcellpros <- data.frame(all_cell %>% filter(comparison == "MgSa_cell"))
MgSa_GOcellpros_list <- as.list(MgSa_GOcellpros)
names(MgSa_GOcellpros_list$Result.gene.list) <- MgSa_GOcellpros_list$Name
MgSaGOlist_cell <- MgSa_GOcellpros_list$Result.gene.list
  #MGPV CELL
MgPv_GOcellpros <- data.frame(all_cell %>% filter(comparison == "MgPv_cell"))
MgPv_GOcellpros_list <- as.list(MgPv_GOcellpros)
names(MgPv_GOcellpros_list$Result.gene.list) <- MgPv_GOcellpros_list$Name
MgPvGOlist_cell <- MgPv_GOcellpros_list$Result.gene.list
  #PVSA CELL
PvSa_GOcellpros <- data.frame(all_cell %>% filter(comparison == "PvSa_cell"))
PvSa_GOcellpros_list <- as.list(PvSa_GOcellpros)
names(PvSa_GOcellpros_list$Result.gene.list) <- PvSa_GOcellpros_list$Name
PvSaGOlist_cell <- PvSa_GOcellpros_list$Result.gene.list
#View(PvSaGOlist_cell)


##################### MOLE PROS #####################
MgSa_GOmolepros <- read.table("MgSa_GOmolecpros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgSa_GOmolepros <- data.frame(MgSa_GOmolepros, comparison = "MgSa_mol")
MgPv_GOmolepros <- read.table("MgPv_GOmolepros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  MgPv_GOmolepros <- data.frame(MgPv_GOmolepros, comparison = "MgPv_mol")
PvSa_GOmolepros <- read.table("PvSa_GOmolepros.tab", sep = "\t",header = TRUE) %>% dplyr :: select(Name, Result.gene.list)
  PvSa_GOmolepros <- data.frame(PvSa_GOmolepros, comparison = "PvSa_mol")
  
all_mol <- rbind(MgSa_GOmolepros,MgPv_GOmolepros, PvSa_GOmolepros)
all_mol$Result.gene.list <- gsub(',',',',all_mol$Result.gene.list,fixed=TRUE)
all_mol$Result.gene.list <- strsplit(all_mol$Result.gene.list, ",")

  #MGSA MOL
MgSa_GOmolepros <- data.frame(all_mol %>% filter(comparison == "MgSa_mol")) 
MgSa_GOmolepros_list <- as.list(MgSa_GOmolepros)
names(MgSa_GOmolepros_list$Result.gene.list) <- MgSa_GOmolepros$Name
MgSaGOlist_mole <- MgSa_GOmolepros_list$Result.gene.list
  #MGPV MOL
MgPv_GOmolepros <- data.frame(all_mol %>% filter(comparison == "MgPv_mol")) 
MgPv_GOmolepros_list <- as.list(MgPv_GOmolepros)
names(MgPv_GOmolepros_list$Result.gene.list) <- MgPv_GOmolepros$Name
MgPvGOlist_mole <- MgPv_GOmolepros_list$Result.gene.list
  #PVSA MOL
PvSa_GOmolepros <- data.frame(all_mol %>% filter(comparison == "PvSa_mol")) 
PvSa_GOmolepros_list <- as.list(PvSa_GOmolepros)
names(PvSa_GOmolepros_list$Result.gene.list) <- PvSa_GOmolepros$Name
PvSaGOlist_mole <- PvSa_GOmolepros_list$Result.gene.list

```


```{r}
####Making SAVAGE-GSEA pathway list from PATHWAY analysis
  #MgSa
MgSaGOpathways <- read.csv("MgSaGO_Pathways.csv") #reading in table from GO tryp database
MgSaGOpathways <- data.frame(MgSaGOpathways) #creating data frame
MgSaGOpathways$Genes <- gsub('|',',',MgSaGOpathways$Genes,fixed=TRUE)#removing the '|' and replacing with ','
MgSaGOpathways$Genes <- strsplit(MgSaGOpathways$Genes, ",")#removing the '|' and replacing with ',' (for some reason it doesnt like the '|')
MgSaGOpathways_list <- as.list(MgSaGOpathways)
names(MgSaGOpathways_list$Genes) <- MgSaGOpathways_list$Pathway
MgSaGO_list <- MgSaGOpathways_list$Genes

####Making GSEA pathway list from PATHWAY analysis 
  #MgPv
MgPvGOpathways <- read.csv("MgPvGO_Pathways.csv")
MgPvGOpathways <- data.frame(MgPvGOpathways)
MgPvGOpathways$Genes <- gsub('|',',',MgPvGOpathways$Genes,fixed=TRUE)
MgPvGOpathways$Genes <- strsplit(MgPvGOpathways$Genes, ",")
MgPvGOpathways_list <- as.list(MgPvGOpathways)
names(MgPvGOpathways_list$Genes) <- MgPvGOpathways_list$Pathway
MgPvGO_list <- MgPvGOpathways_list$Genes
#View(MgPvGO_list)

####Making GSEA pathway list from PATHWAY analysis 
  #PvSa
PvSaGOpathways <- read.csv("PvSaGO_Pathways.csv") #reading in table
PvSaGOpathways <- data.frame(PvSaGOpathways)
PvSaGOpathways$Genes <- gsub('|',',',PvSaGOpathways$Genes,fixed=TRUE) #removing the '|' and replacing with ','
PvSaGOpathways$Genes <- strsplit(PvSaGOpathways$Genes, ',') #removing the '|' and replacing with ','
PvSaGOpathways_list <- as.list(PvSaGOpathways)  
names(PvSaGOpathways_list$Genes) <- PvSaGOpathways_list$Pathway 
PvSaGO_list <- PvSaGOpathways_list$Genes 


```


```{r}

################################GSEA on all SAV DE genes########################################
####need to make list of all 'pathways'
sav_all_list <- c()
  #for MgSa
MgSa_list <- rownames(MgSa)
  #for MgPv
MgPv_list <- rownames(MgPv)
  #for PvSa
PvSa_list <- rownames(PvSa)
 
#creating full list for all savage 
sav_all_list$MgSa <- MgSa_list
sav_all_list$MgPv <- MgPv_list
sav_all_list$PvSa <- PvSa_list

#FGSEA on SAV DE GENES (NO GO)

fgseaRes_all <- fgsea(pathways = as.list(sav_all_list), stats = ranks, minSize = 1,  nperm = 1000)
fgseaRes_all_pos <- fgsea(pathways = as.list(sav_all_list), stats = ranks_log.pos, minSize = 1,  nperm = 1000)
fgseaRes_all_neg <- fgsea(pathways = as.list(sav_all_list), stats = ranks_log.neg, minSize = 1,  nperm = 1000)
 
##########NEG FOLD CHANGE ENRICHMENT#########
plotEnrichment(pathway = sav_all_list$MgSa,
               stats  = ranks_log.neg) + labs(title = "MgSa neg log fold")
plotEnrichment(pathway = sav_all_list$MgPv,
               stats  = ranks_log.neg) + labs(title ="MgPv neg log fold")
plotEnrichment(pathway = sav_all_list$PvSa,
               stats  = ranks_log.neg) + labs(title ="PvSa neg log fold")
##########POS FOLD CHANGE ENRICHMENT#########
plotEnrichment(pathway = sav_all_list$MgSa,
               stats  = ranks_log.pos) + labs(title ="MgSa pos log fold")
plotEnrichment(pathway = sav_all_list$MgPv,
               stats  = ranks_log.pos) + labs(title ="MgPv pos log fold")
plotEnrichment(pathway = sav_all_list$PvSa,
               stats  = ranks_log.pos) + labs(title ="PvSa pos log fold")
##########TOTAL ENRICHMENT##################
plotEnrichment(pathway = sav_all_list$MgSa,
               stats  = ranks ) + labs(title ="MgSa")
plotEnrichment(pathway = sav_all_list$MgPv,
               stats  = ranks ) + labs(title ="MgPv")
plotEnrichment(pathway = sav_all_list$PvSa,
               stats  = ranks ) + labs(title ="PvSa")



```

```{r}
 ########################FGSEA -- SAV GO DATA ENRICHED IN OUR DATA################################
 
#In order to complete GSEA, we must create pathways, or gene lists that help us determine if that particular gene set is enriched in a gene expression data set. Here, we have GO analysis results from three comparisons of savage data, MgSa, MgPv, and PvSa. We will test these pathways on our data, separated in negative and positive log fold change, and ranked by the metric -log(pvalue)*abs(logfold2change). We will preform fgsea on each pathway, for positive, and negative ranked DE genes from our time point 0 to 12hours. 

all_list <- c()
all_list$MgSa_bio <- MgSaGOlist_bio
all_list$MgPv_bio <- MgPvGOlist_bio
all_list$PvSa_bio <- PvSaGOlist_bio
all_list$MgSa_cell <- MgSaGOlist_cell
all_list$MgPv_cell <- MgPvGOlist_cell
all_list$PvSa_cell <- PvSaGOlist_cell
all_list$MgSa_mole <- MgSaGOlist_mole
all_list$MgPv_mole <- MgPvGOlist_mole
all_list$PvSa_mole <- PvSaGOlist_mole

fgsea_all_pos <- c()
fgsea_all_neg <- c()
for (i in 1:9){
fgsea_all_pos[[i]] = fgsea(pathways = as.list(all_list[[i]]), stats = ranks_log.pos, minSize = 15,  nperm = 1000)
fgsea_all_neg[[i]] = fgsea(pathways = as.list(all_list[[i]]), stats = ranks_log.neg, minSize = 15,  nperm = 1000)
}
names(fgsea_all_neg) <- names(all_list)
names(fgsea_all_pos) <- names(all_list)

fgsea(as.list(all_list$MgSa_bio), ranks_log.pos, minSize = 1, nperm = 1000)
#####################MGSA ENRICHMENT########################

########POSLOG###########
plotEnrichment(all_list$MgSa_bio[[head(fgsea_all_pos$MgSa_bio[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgSa_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgSa_cell[[head(fgsea_all_pos$MgSa_cell[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgSa_cell[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgSa_mole[[head(fgsea_all_pos$MgSa_mole[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgSa_mole[order(pval), ], 1)$pathway)
########NEG LOG###########
plotEnrichment(all_list$MgSa_bio[[head(fgsea_all_neg$MgSa_bio[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$MgSa_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgSa_cell[[head(fgsea_all_neg$MgSa_cell[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$MgSa_cell[order(pval), ], 1)$pathway)

#plotEnrichment(all_list$MgSa_mole[[head(fgsea_all_neg$MgSa_mole[order(pval), ], 1)$pathway]],
               #ranks_log.neg) + labs(title=head(fgsea_all_neg$MgSa_mole[order(pval), ], 1)$pathway)

#####################MGPV ENRICHMENT########################

########POS LOG###########
plotEnrichment(all_list$MgPv_bio[[head(fgsea_all_pos$MgPv_bio[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgPv_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgPv_cell[[head(fgsea_all_pos$MgPv_cell[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgPv_cell[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgPv_mole[[head(fgsea_all_pos$MgPv_mole[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$MgPv_mole[order(pval), ], 1)$pathway)
########NEG LOG###########
plotEnrichment(all_list$MgPv_bio[[head(fgsea_all_neg$MgPv_bio[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$MgPv_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgPv_cell[[head(fgsea_all_neg$MgPv_cell[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$MgPv_cell[order(pval), ], 1)$pathway)

plotEnrichment(all_list$MgPv_mole[[head(fgsea_all_neg$MgPv_mole[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$MgPv_mole[order(pval), ], 1)$pathway)
 
#####################PVSA ENRICHMENT########################

########POS LOG###########
plotEnrichment(all_list$PvSa_bio[[head(fgsea_all_pos$PvSa_bio[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$PvSa_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$PvSa_cell[[head(fgsea_all_pos$PvSa_cell[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$PvSa_cell[order(pval), ], 1)$pathway)

plotEnrichment(all_list$PvSa_mole[[head(fgsea_all_pos$PvSa_mole[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos$PvSa_mole[order(pval), ], 1)$pathway)

########NEG LOG###########
plotEnrichment(all_list$PvSa_bio[[head(fgsea_all_neg$PvSa_bio[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$PvSa_bio[order(pval), ], 1)$pathway)

plotEnrichment(all_list$PvSa_cell[[head(fgsea_all_neg$PvSa_cell[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$PvSa_cell[order(pval), ], 1)$pathway)

plotEnrichment(all_list$PvSa_mole[[head(fgsea_all_neg$PvSa_mole[order(pval), ], 1)$pathway]],
               ranks_log.neg) + labs(title=head(fgsea_all_neg$PvSa_mole[order(pval), ], 1)$pathway)
```

```{r}
############################FGSEA ON 2019 CLUSTERS#############################
#In order to answer the biological questions of our data, we need to take statisitcal steps to gather information in order to prove, or answer the hypothesis at hand. Clustering our data is an example of this method. Here we are taking the fuzzy cluster GO analysis results, and preparing them for GSEA pipeline. 

#Reading in huge file with all GO terms of cluster.s 
fuzzyClust <- read.table("fuzzyclusters_GO.tab") %>% dplyr :: select(Name, Result.gene.list, Cluster,Type)
fuzzyClust$Result.gene.list <- gsub(',',',',fuzzyClust$Result.gene.list,fixed=TRUE)
fuzzyClust$Result.gene.list <- strsplit(fuzzyClust$Result.gene.list, ",")


#cluster1 <- data.frame(GO[1]) %>%dplyr::select(Cluster1.Name,Cluster1.Result.gene.list,Cluster1.Type) 
#cluster1$Cluster1.Result.gene.list <- gsub(',',',',cluster1$Cluster1.Result.gene.list,fixed=TRUE)
#cluster1$Cluster1.Result.gene.list <- strsplit(cluster1$Cluster1.Result.gene.list, ",")

clust1_bio <- fuzzyClust %>% filter(Type == "Biological Process", Cluster == 1)
clust1_bio_list <- as.list(clust1_bio)
names(clust1_bio_list$Result.gene.list) <- clust1_bio_list$Name
clust1_bio <- clust1_bio_list$Result.gene.list
View(clust1_bio)

clust1_cell <- fuzzyClust %>% filter(Type == "Cellular Component", Cluster == 1)
clust1_cell_list <- as.list(clust1_cell)
names(clust1_cell_list$Result.gene.list) <- clust1_cell_list$Name
clust1_cell <- clust1_cell_list$Result.gene.list
#View(fuzzylist_1_cell)

clust1_mol <- fuzzyClust %>% filter(Type == "Molecular Function",Cluster == 1)
clust1_mol_list <- as.list(clust1_mol)
names(clust1_mol_list$Result.gene.list) <- clust1_mol_list$Name
clust1_mol <- clust1_mol_list$Result.gene.list

#########CLUSTER 2######### 
 
clust2_bio <- fuzzyClust %>% filter(Type == "Biological Process", Cluster == 2)
clust2_bio_list <- as.list(clust2_bio)
names(clust2_bio_list$Result.gene.list) <- clust2_bio_list$Name
clust2_bio <- clust2_bio_list$Result.gene.list
#View(clust2_bio)

clust2_cell <- fuzzyClust %>% filter(Type == "Cellular Component", Cluster == 2)
clust2_cell_list <- as.list(clust2_cell)
names(clust2_cell_list$Result.gene.list) <- clust2_cell_list$Name
clust2_cell <- clust2_cell_list$Result.gene.list
View(fuzzylist_2_cell)

#clust2_mol <- cluster2 %>% filter(Cluster2.Type == "Molecular Function")
#clust2_mol_list <- as.list(clust2_mol)
#names(clust2_mol_list$Cluster2.Result.gene.list) <- clust2_mol_list$Cluster2.Name
#fuzzylist_2_mol <- clust2_mol_list$Cluster2.Result.gene.list
#View(fuzzylist_2_mol)

#cluster3 <- data.frame(GO[3])%>%dplyr::select(Cluster3.Name,Cluster3.Result.gene.list,Cluster3.Type) 
#clust3_bio <- cluster3 %>% filter(Cluster3.Type == "Biological Process")
#clust3_cell <- cluster3 %>% filter(Cluster3.Type == "Cellular Component")
#clust3_mol <- cluster3 %>% filter(Cluster3.Type == "Molecular Function")

clust4_bio <- fuzzyClust %>% filter(Type == "Biological Process", Cluster == 4)
clust4_bio_list <- as.list(clust4_bio)
names(clust4_bio_list$Result.gene.list) <- clust4_bio_list$Name
clust4_bio <- clust4_bio_list$Result.gene.list
#View(fuzzylist_4_bio)

clust4_cell <- fuzzyClust %>% filter(Type == "Cellular Component", Cluster == 4)
clust4_cell_list <- as.list(clust4_cell)
names(clust4_cell_list$Result.gene.list) <- clust4_cell_list$Name
clust4_cell <- clust4_cell_list$Result.gene.list
#View(fuzzylist_4_cell)

clust4_mol <- fuzzyClust %>% filter(Type == "Molecular Function", Cluster == 4)
clust4_mol_list <- as.list(clust4_mol)
names(clust4_mol_list$Result.gene.list) <- clust4_mol_list$Name
clust4_mol <- clust4_mol_list$Result.gene.list
View(clust4_mol)
#cluster5 <- data.frame(GO[5])%>%dplyr::select(Cluster5.Name,Cluster5.Result.gene.list,Cluster5.Type) 
#clust5_bio <- cluster5 %>% filter(Cluster5.Type == "Biological Process")
#clust5_cell <- cluster5 %>% filter(Cluster5.Type == "Cellular Component")
#clust5_mol <- cluster5 %>% filter(Cluster5.Type == "Molecular Function")
 

#clust6_bio <- cluster6 %>% filter(Cluster6.Type == "Biological Process")
clust6_cell <- fuzzyClust %>% filter(Type == "Cellular Component", Cluster == 6) 
clust6_cell_list <- as.list(clust6_cell)
names(clust6_cell_list$Result.gene.list) <- clust6_cell_list$Name
clust6_cell <- clust6_cell_list$Result.gene.list

#making Result.gene.list entry into list 
 
all_fuzzy <- c()
all_fuzzy$bio1 <-  clust1_bio 
all_fuzzy$bio2 <- clust2_bio
all_fuzzy$bio4 <- clust4_bio
#all_fuzzy$bio6 <- clust6_bio

all_fuzzy$cell1 <- clust1_cell
all_fuzzy$cell2 <- clust2_cell
all_fuzzy$cell4 <- clust4_cell
all_fuzzy$cell6 <- clust6_cell

all_fuzzy$mol1 <- clust1_mol
#all_fuzzy$mol2 <- clust2_mol
all_fuzzy$mol4 <- clust4_mol
#all_fuzzy$mol6 <- clust6_mol

View(all_fuzzy)
#library(fgsea)
 
fgsea(clust1_bio, ranks_MgSa, nperm = 1000)

exampleRanks
fgsea_mgsa_fuzzy <- c()
fgsea_mgpv_fuzzy <- c()
fgsea_pvsa_fuzzy <- c()

BiocManager::install("BiocParallel")

library(BiocParallel)
bplapply(c(pathways = as.list(all_fuzzy$cell6), stats = ranks_MgSa, minSize = 1,  nperm = 1000), fgsea, BPPARAM = SnowParam())
View(all_fuzzy$bio1)
for (i in 1:9){
fgsea_mgsa_fuzzy[[i]] = fgsea(pathways = as.list(all_fuzzy[[i]]), stats = ranks_MgSa, minSize = 1,  nperm = 100)
fgsea_mgpv_fuzzy[[i]] = fgsea(pathways = as.list(all_fuzzy[[i]]), stats = ranks_MgPv, minSize = 1,  nperm = 100)
fgsea_pvsa_fuzzy[[i]] = fgsea(pathways = as.list(all_fuzzy[[i]]), stats = ranks_PvSa, minSize = 1,  nperm = 100)
}
ranks_MgSa
names(fgsea_mgsa_fuzzy) <- names(all_fuzzy)
names(fgsea_mgpv_fuzzy) <- names(all_fuzzy)
names(fgsea_pvsa_fuzzy) <- names(all_fuzzy)
#View(fgsea_mgsa_fuzzy)
#View(fgsea_mgpv_fuzzy)
#View(fgsea_pvsa_fuzzy)

#fgsea(as.list(all_fuzzy$bio1), ranks_MgSa, minSize = 15,  nperm = 100)

#head(ranks_MgSa)
#head(exampleRanks)
#all_fuzzy$cell1
#length(ranks_PvSa)
c <- intersect((clust1_bio$`movement of cell or subcellular component`), names(ranks_MgPv))
length(c)
length(clust1_bio$`movement of cell or subcellular component`)
```

```{r}
############clust 1###############
plotEnrichment(all_fuzzy$cell1[[head(fgsea_all_fuzzy$cell1[order(pval), ], 1)$pathway]],
               ranks) + labs(title=head(fgsea_all_fuzzy$cell1[order(pval), ], 1)$pathway)

plotEnrichment(all_fuzzy$mol1[[head(fgsea_all_pos_fuzzy$mol1[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos_fuzzy$mol1[order(pval), ], 1)$pathway)

############clust 2###############
plotEnrichment(all_fuzzy$bio2[[head(fgsea_all_fuzzy$bio2[order(pval), ], 1)$pathway]],
               ranks) + labs(title=head(fgsea_all_fuzzy$bio2[order(pval), ], 1)$pathway)
plotEnrichment(all_fuzzy$cell2[[head(fgsea_all_pos_fuzzy$cell2[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos_fuzzy$cell2[order(pval), ], 1)$pathway)

############clust 4###############
plotEnrichment(all_fuzzy$cell4[[head(fgsea_all_pos_fuzzy$cell4[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos_fuzzy$cell4[order(pval), ], 1)$pathway)

plotEnrichment(all_fuzzy$mol4[[head(fgsea_all_pos_fuzzy$mol4[order(pval), ], 1)$pathway]],
               ranks_log.pos) + labs(title=head(fgsea_all_pos_fuzzy$mol4[order(pval), ], 1)$pathway)

plotEnrichment(all_fuzzy$bio4[[head(fgsea_all_fuzzy$bio4[order(pval), ], 1)$pathway]],
               ranks) + labs(title=head(fgsea_all_fuzzy$bio4[order(pval), ], 1)$pathway)


#View(fgsea_all_pos_fuzzy$cell4)

c <- intersect(clust1_bio$'microtubule-based movement', as.character(rownames(MgSa)))
length(c)
```
