---
title: "E. coli Technical Report"
author: "Annie Cohen"
date: "7/22/2019"
output:
  pdf_document: default
  html_document: default
---



## Table of Contents
**1.** Background on *E. coli*  
**2.** RNA-Seq  
**3.** Our Data  
**4.** Normalization  
**5.** Determine Differentially Expressed Genes 
**5.** Visualize Genes with Shiny App  
**6.** Clustering

# 1 Background on *E. coli*

$~$

# 2 RNA-Seq

$~$

# 3 Our data
The data that we recieved from Professor Stoebel and his lab is in the format of a count matrix, with 37 columns. The column names specify the strain, timepoint, treatment, and replicate of the sample. "JH_01" or "JH_02" specifies the strain of *E. coli*, where "01" is wildtype and "02" is dRpoS. "A", "B", and "C" denotes the biological replicate and the number following it indicates the time point, "01" is 0, "02" is 30, "03" is 60, "04" is 90, "05" is 120, and "06" is 150 minutes. The original data came with a column called "Geneid". In order to normalize our data, we had to assign the column as the rownames and then delete the column.
$~$

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
library(ggplot2)
library(tidyr)
library(shiny)
library(dplyr)
library(DESeq2)
library(readr)
library(ImpulseDE2)
library(maSigPro)
library(VennDiagram)
```

```{r}
# import data
ec_rawcounts <- read.table("LB_Time_Course_GCA_000005845.2_ASM584v2_genomic_counts.txt", header = TRUE)

# assign Geneid as rownames and delete variable
rownames(ec_rawcounts) <- ec_rawcounts$Geneid
ec_rawcounts <- ec_rawcounts %>% dplyr::select(-Geneid)
```
$~$

# 4 Normalization

We tested out many normalization methods including DESeq2 and edgeR. DESeq2 is a normalization and differential expression analysis method that divides counts by size factors that are calculated as the median of the ratios of read counts to the geometric mean. EdgeR employs the trimmed mean method (TMM), where gene expression is adjusted by the fold change and relative expression level to its sample (Ciaran et. al). Through a comparison of size factors, we determined that the two methods produced very similar normalized counts. We ultimately decided to use DESeq2 because the algorithm was easier for our group to understand.
{insert image of size factor plot}
$~$

```{r}
# run DESeq2
ec_coldata <- data.frame(names = colnames(ec_rawcounts)) %>%
  separate(names, c("treat", "rep_time"), "_") %>%
  mutate(rep = ifelse(substr(rep_time,start=1,stop = 1) == "A", 1, 
                        ifelse(substr(rep_time,start=1,stop = 1) == "B", 2, 3))) %>%
  mutate(time = ifelse(substr(rep_time,start=3,stop = 3) == "1", 0,
                      ifelse(substr(rep_time,start=3,stop = 3) == "2", 30,
                             ifelse(substr(rep_time,start=3,stop = 3) == "3", 60,
                                    ifelse(substr(rep_time,start=3,stop = 3) == "4", 90,
                                           ifelse(substr(rep_time,start=3,stop = 3) == "5", 120, 150)))))) %>%
  mutate(treat = ifelse(treat == "JH01", "WT","dRpoS")) %>%
  mutate(timetreat = paste(treat,time,sep = "")) %>%
  select(-rep_time)
rownames(ec_coldata) <- colnames(ec_rawcounts)
```

```{r, cache=TRUE}
ec_dds <- DESeqDataSetFromMatrix(ec_rawcounts, colData = ec_coldata, design = ~ timetreat)
ec_dds_de <- DESeq(ec_dds)
ec_normcounts <- as.data.frame(counts(ec_dds_de, normalized=TRUE))
```
$~$

# 5 Determine Differentially Expressed Genes
## 5.1 DESeq2

```{r}
EC_sig_genes <- c()

# determine the genes that are differentially expressed between WT0 and WT150
de_results <- results(ec_dds_de, contrast = c("timetreat", "WT150","WT0"))
resSig <- de_results[ which(de_results$padj < 0.01 ), ]
resSig$Geneid <- rownames(resSig)
ind <- match(rownames(ec_normcounts),resSig$Geneid)
ec_normcounts$sig <- resSig$Geneid[ind]
EC_sig_genes$WT_compare <- ec_normcounts %>%
 filter(!is.na(ec_normcounts$sig))


# determine the genes that are differentially expressed between WT150 and dRpoS150
de_results <- results(ec_dds_de, contrast = c("timetreat", "WT150","dRpoS150"))
resSig <- de_results[ which(de_results$padj < 0.01 ), ]
resSig$Geneid <- rownames(resSig)
ind <- match(rownames(ec_normcounts),resSig$Geneid)
ec_normcounts$sig <- resSig$Geneid[ind]
EC_sig_genes$treat_compare <- ec_normcounts %>%
 filter(!is.na(ec_normcounts$sig))

# find the intersection of the two lists of differentially expressed genes
EC_sig_genes <- intersect(EC_sig_genes$WT_compare$sig, EC_sig_genes$treat_compare$sig)
```

## 5.2 ImpulseDE2

```{r, eval=FALSE}
ec_design <- data.frame(Sample=colnames(ec_rawcounts),
                        Condition=c(rep("case",6), rep("control",6), rep("case",6), rep("control",6), rep("case",6), rep("control",6)), 
                        Time=rep(c(0,30,60,90,120,150),6))
```

```{r, cache=TRUE, eval = FALSE}
impulse_ecoli <- ImpulseDE2::runImpulseDE2(as.matrix(ec_rawcounts), ec_design, boolCaseCtrl=TRUE, vecConfounders=NULL, scaQThres = 0.01, scaNProc = 8, boolIdentifyTransients = TRUE)
```

## 5.3 MaSigPro

```{r, eval = FALSE}
# set up design
Time <- rep(c(0,30,60,90,120,150),6)
Replicate <- rep(1:3, 12)
Control <- c(rep(0, 6), rep(1, 6), rep(0, 6), rep(1, 6), rep(0, 6), rep(1,6))
d_Rpos <- c(rep(1, 6), rep(0, 6), rep(1, 6), rep(0, 6), rep(1, 6), rep(0, 6))
ec_design <- cbind(Time, Replicate, Control, d_Rpos)
rownames(ec_design) <- colnames(ec_rawcounts)

# run maSigPro on DESeq2 Normalized Counts, theta=10, Q=0.01
ec_normcounts <- as.data.frame(counts(ec_dds_de, normalized=TRUE))
ec_design <- make.design.matrix(ec_design, degree=5)
```

```{r, cache=TRUE, eval=FALSE}
# ID DEGs
fits <- p.vector(ec_normcounts, ec_design, Q=0.01, MT.adjust = "BH", min.obs=6, counts=TRUE)
#variable selection procedure to ID variables for each gene
tsep <- T.fit(fits, step.method = "backward", alfa=0.01)

# Get sig genes between delta_RpoS and Control
sigs <- get.siggenes(tsep, rsq = 0.6, vars = "groups")
maSigProGenes<-sigs$summary$d_RposvsControl
```

$~$

## 5.4 Find the Intersection

```{r}
# Intersect each gene list with another method to calculate gene list of DEGs found in 2 < methods

DS_IM <- distinct(data.frame("Geneid" = intersect(EC_sig_genes, impulse_ecoli$vecDEGenes)))
IM_MA <- distinct(data.frame("Geneid" = intersect(impulse_ecoli$vecDEGenes, maSigProGenes)))
MA_DS <- distinct(data.frame("Geneid" = intersect(maSigProGenes, EC_sig_genes)))

EC_DEGs <- distinct(data.frame("Geneid" = c(DS_IM$Geneid, IM_MA$Geneid, MA_DS$Geneid)))
```


This is a visualization of the overlap between these three methods. Our `EC_DEGs` list is a compilation of genes that overlap between two or more methods:

```{r, echo=FALSE, cache=TRUE}
# calculate intersection of all three methods' gene lists
all_intersect <- intersect(intersect(EC_sig_genes, impulse_ecoli$vecDEGenes), maSigProGenes)

# calculate the size of each gene list including the intersection lists
DS <- length(EC_sig_genes)
IM <- length(impulse_ecoli$vecDEGenes)
MA <- length(maSigProGenes)
all_length <- length(all_intersect)
DS_IM_length <- length(DS_IM$Geneid)
IM_MA_length <- length(IM_MA$Geneid)
MA_DS_length <- length(MA_DS$Geneid)

# plot venn diagram
grid.newpage()
draw.triple.venn(area1 = DS, area2 = IM, area3 = MA, n12= DS_IM_length, n13 = MA_DS_length, n23 = IM_MA_length, n123 = all_length, category = c("DESeq2", "ImpulseDE2", "MaSigPro"), fill = c("lightblue", "khaki1", "pink"), lty = "blank")
```